sgkit/io/vcf/vcf_writer.py

```diff
@@ -60,36 +106,37 @@ def write_vcf(
         python -c 'import sys; from sgkit.io.vcf import zarr_to_vcf; zarr_to_vcf("in.zarr", sys.stdout)'
             | bgzip > out.vcf.gz
 
-    Warnings
-    --------
-    This function requires the dataset to have a ``vcf_header`` attribute
-    containing the VCF header. VCF files converted to Zarr using :func:`vcf_to_zarr`
-    will contain this attribute, but datasets loaded from other sources will not.
-
     Parameters
     ----------
     input
         Dataset to convert to VCF.
     output
         A path or text file object that the output VCF should be written to.
+    vcf_header
+        The VCF header to use (including the line starting with ``#CHROM``). If omitted, then
+        a header will be generated from the dataset fields.
     """
 
     with ExitStack() as stack:
         if isinstance(output, str) or isinstance(output, Path):
             output = stack.enter_context(open(output, mode="w"))
 
-        header_str = input.attrs["vcf_header"]
+        if vcf_header is None:
```

Comment:
This feels like a breaking change - shouldn't we at least warn the user if there is a ``vcf_header`` attribute? Would it not make sense to use the ``vcf_header`` attribute if we already have one? I.e., parse the header (if it exists) and check whether all the fields match up, and generate any new ones as necessary?

Reply:
Yes, it is a breaking change (mentioned on #1020), and a warning is probably a good idea.

VCF header management is tricky though, and merging two headers even more so. What would we do if the version of the VCF header in the attribute is older than the one we are using to write? VCF 4.3 had lots of changes in things like header name validation - basically making things stricter - so it wouldn't be very safe to use a line from the original header since it might not be valid in the later version. (To take one example, PEDIGREE fields now require an ID.)

Reply:
I agree it's very tricky. Maybe I'm misunderstanding though - what happens in the case where we do something like

- Read in a VCF to dataset
- Compute something which changes the dataset (say, phase the data, just for example)
- Export back out to VCF

Do we lose all header information beyond the fields and types, like comments and provenance information from the original?

I agree parsing out the header and trying to understand every bit of it would be very difficult, but you could do something simpler like assuming that if header information exists for a given field, it's correct.

Reply:
To give an example, here is a diff of the output header (generated by the code in this PR) with the input header for the sample.vcf in this repo. (The lines are sorted to make it easier to compare - normally they are not in this order of course.)

<details>

```
E             + ##ALT=<ID=CNV,Description="Copy number variable region">
E             + ##ALT=<ID=DEL:ME:ALU,Description="Deletion of ALU element">
E               ##FILTER=<ID=PASS,Description="All filters passed">
E             - ##FILTER=<ID=q10,Description="">
E             + ##FILTER=<ID=q10,Description="Quality below 10">
E             ?                               ++++++++++++++++
E             - ##FILTER=<ID=s50,Description="">
E             + ##FILTER=<ID=s50,Description="Less than 50% of samples have data">
E               ##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
E               ##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
E               ##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
E               ##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality">
E               ##INFO=<ID=AA,Number=1,Type=String,Description="Ancestral Allele">
E             - ##INFO=<ID=AC,Number=2,Type=Integer,Description="Allele count in genotypes, for each ALT allele, in the same order as listed">
E             ?                      ^
E             + ##INFO=<ID=AC,Number=.,Type=Integer,Description="Allele count in genotypes, for each ALT allele, in the same order as listed">
E             ?                      ^
E             - ##INFO=<ID=AF,Number=2,Type=Float,Description="Allele Frequency">
E             ?                      ^
E             + ##INFO=<ID=AF,Number=.,Type=Float,Description="Allele Frequency">
E             ?                      ^
E               ##INFO=<ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes">
E               ##INFO=<ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129">
E               ##INFO=<ID=DP,Number=1,Type=Integer,Description="Total Depth">
E               ##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership">
E               ##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of Samples With Data">
E               ##contig=<ID=19>
E               ##contig=<ID=20>
E               ##contig=<ID=X>
E             + ##fileDate=20090805
E             - ##fileformat=VCFv4.3
E             ?                    ^
E             + ##fileformat=VCFv4.0
E             ?                    ^
E             - ##source=sgkit-0.6.1.dev2+gcc728043
E             + ##phasing=partial
E             + ##reference=1000GenomesPilot-NCBI36
```

</details>

Summary of differences:
* `fileformat` version numbers
* `source`
* Extra unstructured (key/value) fields in the original:
  * `fileDate`
  * `phasing`
  * `reference`
* Extra structured fields in the original:
  * `ALT`
* Missing description in `FILTER`
* More specific `Number` (e.g. `.` in original, `2` in output)

I've run this on other test files in the repo, and this difference crops up too:
* Missing `assembly` field in `contig` (related to #464)

Reply:
> I agree parsing out the header and trying to understand every bit of it would be very difficult, but you could do something simpler like assuming that if header information exists for a given field, it's correct.

Good idea - I'm investigating this approach...